services/src/str_util.cpp

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2018 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "str_util.h"
+
+namespace particle {
+
+char* escape(const char* src, size_t srcSize, const char* spec, size_t specSize, char esc, char* dest, size_t destSize) {
+    size_t i = 0;
+    size_t j = 0;
+    while (i < srcSize && j < destSize) {
+        const char c = src[i++];
+        if (memchr(spec, (unsigned char)c, specSize) != nullptr) {
+            if (j == destSize - 1) {
+                break; // Avoid generating an invalid escaped string
```

Comment:
How can the caller know the buffer was too small?

Reply:
There's no way to know that. We can make this function fill the entire output buffer and return the number of characters required to store the complete escaped string, which can be larger than the size of the output buffer – similarly to how `printf()` family of functions work. I just wasn't sure if it's necessary, given that the caller always knows the size of the buffer in the worst case (`strlen(src) * 2 + 1`)